File: daily_python/2126_Destroying_Asteroids.py
import os
import sys
from typing import List

class Solution:
    def asteroidsDestroyed(self, mass: int, asteroids: List[int]) -> bool:
        m = [100001] * 33
        s = [0] * 33
        for a in asteroids:
            k = a.bit_length()
            m[k] = min(m[k], a)
            s[k] += a
        for i in range(33):
            if s[i] == 0:
                continue
            if mass < m[i]:
                return False
            mass += s[i]
        return True

def readDataSet(filename):
    dataset = []
    with open(filename, 'r') as file:
        content = file.read().strip()
        blocks = content.split('\n\n')
        for block in blocks:
            lines = block.split('\n')
            mass = int(lines[0].split('=')[1].strip()[:-1])
            asteroids = list(map(int, lines[1].split('=')[1].strip()[1:-2].split(',')))
            dataset.append((mass, asteroids))
    return dataset

if __name__ == '__main__':
    if len(sys.argv) == 1:
        filename = os.path.splitext(__file__)[0] + '.txt'
    else:
        filename = sys.argv[1]
    dataset = readDataSet(filename)
    results = []
    solution = Solution()
    for mass, asteroids in dataset:
        results.append(solution.asteroidsDestroyed(mass, asteroids))
    for index, result in enumerate(results):
        print(f'Example {index + 1} : {result}')

File: daily/2026/2126_Destroying_Asteroids.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    bool asteroidsDestroyed(int mass, vector<int>& asteroids) {
        long long m = mass;
        int mins[33];
        long long sums[33];
        fill(mins, mins + 33, INT_MAX);
        fill(sums, sums + 33, 0);
        for (int a : asteroids){
            int k = 32 - __builtin_clz(a);
            mins[k] = min(mins[k], a);
            sums[k] += a;
        }
        for (int i = 0; i < 33; i++){
            if (sums[i] == 0) continue;
            if (m < mins[i]) return false;
            m += sums[i];
        }
        return true;
    }
};

int main(){
    int mass = 10;
    vector<int> asteroids = {3,9,19,5,21};
    // int mass = 5;
    // vector<int> asteroids = {4,9,23,4};
    Solution S;
    cout << S.asteroidsDestroyed(mass, asteroids) << endl;
    return 0;
}